ddl: Add `sql_require_primary_key` sysvar

joycse06 opened the pull request:
<!--

Thank you for contributing to TiDB!

PR Title Format:
1. pkg [, pkg2, pkg3]: what's changed
2. *: what's changed

-->

### What problem does this PR solve?
<!--

Please create an issue first to describe the problem.

There MUST be one line starting with "Issue Number:  " and 
linking the relevant issues via the "close" or "ref".

For more info, check https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/contribute-code.html#referring-to-an-issue.

-->

Issue Number: close https://github.com/pingcap/tidb/issues/28544

Problem Summary: 

This is a feature from MySQL 8.0:
https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_sql_require_primary_key

It allows you to enforce tables have a primary key. It has more meaning in MySQL because row-based replication benefits from it, and InnoDB is less efficient when using its internal primary key. But it's still useful.

### What is changed and how it works?

`sql_require_primary_key` sysvar has been added. When enabled it will throw an error if a create table ddl doesn't have a primary key.

### Check List

Tests <!-- At least one of them must be included. -->

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

<!-- compatibility change, improvement, bugfix, and new feature need a release note -->

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

dveeden commented:
I don't think this should be allowed:
```
sql> SET SESSION sql_require_primary_key=ON;
Query OK, 0 rows affected (0.0007 sec)

sql> SHOW CREATE TABLE t4\G
*************************** 1. row ***************************
       Table: t4
Create Table: CREATE TABLE `t4` (
  `id` int(11) NOT NULL,
  `c1` int(11) DEFAULT NULL,
  PRIMARY KEY (`id`) /*T![clustered_index] NONCLUSTERED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
1 row in set (0.0010 sec)

sql> ALTER TABLE t4 DROP PRIMARY KEY;
Query OK, 0 rows affected (0.0222 sec)
```

7yyo commented:
https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_sql_require_primary_key 

It doesn't just work on create table

morgo commented:
Getting a linter error:
```
[2022-07-12T11:52:36.931Z]   ✘  https://revive.run/r#var-naming  const SqlRequirePrimaryKey should be SQLRequirePrimaryKey  
[2022-07-12T11:52:36.932Z]   sessionctx/variable/sysvar.go:1987:2
```

You may also need to run `make errdoc` to update the errors file.

joycse06 commented:
>Not sure if this should apply to temp tables:

MySQL 8 applies it to temporary tables as well.

[From the doc](https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_sql_require_primary_key:~:text=sql_require_primary_key%20applies%20to%20both%20base%20tables%20and%20TEMPORARY%20tables):

> sql_require_primary_key applies to both base tables and TEMPORARY tables

joycse06 commented:
I wasn't using the correct error code but a similar one. I have now added the actual error code. Have also added check to stop dropping Primary Key indexes (non-clustered one's) when this is set.

```
tidb> set sql_require_primary_key = 1;
Query OK, 0 rows affected (0.00 sec)

tidb> SHOW VARIABLES LIKE 'sql_require_primary_key';
+-------------------------+-------+
| Variable_name           | Value |
+-------------------------+-------+
| sql_require_primary_key | ON    |
+-------------------------+-------+
1 row in set (0.00 sec)

tidb> CREATE TABLE `t4` (`id` int(11) NOT NULL, `c1` int(11) DEFAULT NULL);
ERROR 3750 (HY000): Unable to create or change a table without a primary key, when the system variable 'sql_require_primary_key' is set. Add a primary key to the table or unset this variable to avoid this message. Note that tables without a primary key can cause performance problems in row-based replication, so please consult your DBA before changing this setting.
tidb> CREATE TABLE `t4` (`id` int(11) NOT NULL, `c1` int(11) DEFAULT NULL, PRIMARY KEY (`id`));
Query OK, 0 rows affected (0.01 sec)

tidb> ALTER TABLE t4 DROP COLUMN id;
ERROR 8200 (HY000): Unsupported drop integer primary key
tidb> ALTER TABLE t4 DROP PRIMARY KEY;
ERROR 8200 (HY000): Unsupported drop primary key when the table's pkIsHandle is true
tidb> DROP TABLE t4;
Query OK, 0 rows affected (0.01 sec)

tidb> CREATE TABLE `t4` (`id` int(11) NOT NULL, `c1` int(11) DEFAULT NULL, PRIMARY KEY (`id`) NONCLUSTERED);
Query OK, 0 rows affected (0.00 sec)

tidb> ALTER TABLE t4 DROP COLUMN id;
ERROR 8200 (HY000): can't drop column id with composite index covered or Primary Key covered now
tidb> ALTER TABLE t4 DROP PRIMARY KEY;
ERROR 3750 (HY000): Unable to create or change a table without a primary key, when the system variable 'sql_require_primary_key' is set. Add a primary key to the table or unset this variable to avoid this message. Note that tables without a primary key can cause performance problems in row-based replication, so please consult your DBA before changing this setting.
```

For some cases TiDB is returning `ERROR 8200` instead of  `ERROR 3750` for thats coming from existing code, I am not sure if I should change those.

joycse06 commented on errors.toml:
```diff
@@ -2231,6 +2231,11 @@ error = '''
 
+["schema:3750"]
+error = '''
+Unable to create or change a table without a primary key, when the system variable 'sql_require_primary_key' is set. Add a primary key to the table or unset this variable to avoid this message. Note that tables without a primary key can cause performance problems in row-based replication, so please consult your DBA before changing this setting.
```
This is the same error mysql returns, copied from [here](https://dev.mysql.com/doc/mysql-errors/8.0/en/server-error-reference.html#:~:text=added%20in%208.0.13.-,Error%20number%3A%203750,-%3B%20Symbol%3A%20ER_TABLE_WITHOUT_PK)

dveeden commented:
> ERROR 3750 (HY000): Unable to create or change a table without a primary key, when the system variable 'sql_require_primary_key' is set. Add a primary key to the table or unset this variable to avoid this message. Note that tables without a primary key can cause performance problems in row-based replication, so please consult your DBA before changing this setting.

The "...row-based replication..." part is a bit weird in a TiDB context, we could change this to:

> ERROR 3750 (HY000): Unable to create or change a table without a primary key, when the system variable 'sql_require_primary_key' is set. Add a primary key to the table or unset this variable to avoid this message.

But keeping this message identical to MySQL has my personal preference.

dveeden commented:
/cc @7yyo

morgo commented on ddl/ddl_api.go:
```diff
@@ -1998,6 +1998,11 @@ func checkTableInfoValidWithStmt(ctx sessionctx.Context, tbInfo *model.TableInfo
 	}
+
+	// Check if table has a primary key if required.
+	if ctx.GetSessionVars().PrimaryKeyRequired && len(tbInfo.GetPkName().String()) == 0 {
+		return infoschema.ErrTableWithoutPrimaryKey
+	}
```
I took a look at the internal tables, and many don't have primary keys. Until we can clean it up, we should make this only apply to user-sessions:
```suggestion
	if !ctx.GetSessionVars().InRestrictedSQL && ctx.GetSessionVars().PrimaryKeyRequired && len(tbInfo.GetPkName().String()) == 0 {
		return infoschema.ErrTableWithoutPrimaryKey
	}
```

dveeden replied:
I found this:
```
sql> SET SESSION group_concat_max_len=120;
Query OK, 0 rows affected (0.0003 sec)

sql> SELECT t.TABLE_SCHEMA,COUNT(*),GROUP_CONCAT(t.TABLE_NAME) FROM information_schema.TABLES t LEFT JOIN information_schema.KEY_COLUMN_USAGE kcu ON kcu.TABLE_SCHEMA=t.TABLE_SCHEMA AND kcu.TABLE_NAME=t.TABLE_NAME AND kcu.CONSTRAINT_NAME='PRIMARY' WHERE kcu.CONSTRAINT_NAME IS NULL GROUP BY TABLE_SCHEMA;
+--------------------+----------+--------------------------------------------------------------------------------------------------------------------------+
| TABLE_SCHEMA       | COUNT(*) | GROUP_CONCAT(t.TABLE_NAME)                                                                                               |
+--------------------+----------+--------------------------------------------------------------------------------------------------------------------------+
| PERFORMANCE_SCHEMA |       30 | global_status,session_status,setup_actors,setup_objects,setup_instruments,setup_consumers,events_statements_current,even |
| METRICS_SCHEMA     |      639 | tikv_grpc_avg_req_batch_size,tikv_region_written_bytes,tikv_scheduler_keys_written_avg,tikv_lock_manager_detect_error_to |
| mysql              |       13 | stats_meta,stats_histograms,stats_buckets,gc_delete_range,gc_delete_range_done,stats_feedback,bind_info,stats_top_n,expr |
| INFORMATION_SCHEMA |       76 | OPTIMIZER_TRACE,TIDB_HOT_REGIONS_HISTORY,INSPECTION_SUMMARY,TIFLASH_SEGMENTS,PLUGINS,VIEWS,TIFLASH_REPLICA,CLUSTER_SYSTE |
+--------------------+----------+--------------------------------------------------------------------------------------------------------------------------+
4 rows in set, 1 warning (0.0058 sec)
Warning (code 1260): Some rows were cut by GROUPCONCAT(Column#3)
```
Would we need to add PK's to all of these or just the ones in the `mysql` schema?

morgo commented on ddl/ddl_api.go:
```diff
@@ -6090,6 +6095,10 @@ func (d *ddl) dropIndex(ctx sessionctx.Context, ti ast.Ident, indexName model.CI
 		return err
 	}
 
+	if ctx.GetSessionVars().PrimaryKeyRequired && isPK {
+		return infoschema.ErrTableWithoutPrimaryKey
+	}
```
Same here:
```suggestion
	if !ctx.GetSessionVars().InRestrictedSQL && ctx.GetSessionVars().PrimaryKeyRequired && isPK {
		return infoschema.ErrTableWithoutPrimaryKey
	}
```

morgo commented:
LGTM, please resolve conflicts.

dveeden commented:
Note that DM doesn't use real PKs either:
```
sql> show create table dm_meta.`demo-mysql8-tidb_loader_checkpoint`\G
*************************** 1. row ***************************
       Table: demo-mysql8-tidb_loader_checkpoint
Create Table: CREATE TABLE `demo-mysql8-tidb_loader_checkpoint` (
  `id` char(32) NOT NULL,
  `filename` varchar(255) NOT NULL,
  `cp_schema` varchar(128) NOT NULL,
  `cp_table` varchar(128) NOT NULL,
  `offset` bigint(20) NOT NULL,
  `end_pos` bigint(20) NOT NULL,
  `create_time` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
  `update_time` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  UNIQUE KEY `uk_id_f` (`id`,`filename`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
1 row in set (0.0005 sec)

sql> show create table dm_meta.`demo-mysql8-tidb_syncer_checkpoint`\G
*************************** 1. row ***************************
       Table: demo-mysql8-tidb_syncer_checkpoint
Create Table: CREATE TABLE `demo-mysql8-tidb_syncer_checkpoint` (
  `id` varchar(32) NOT NULL,
  `cp_schema` varchar(128) NOT NULL,
  `cp_table` varchar(128) NOT NULL,
  `binlog_name` varchar(128) DEFAULT NULL,
  `binlog_pos` int(10) unsigned DEFAULT NULL,
  `binlog_gtid` text DEFAULT NULL,
  `exit_safe_binlog_name` varchar(128) DEFAULT '',
  `exit_safe_binlog_pos` int(10) unsigned DEFAULT '0',
  `exit_safe_binlog_gtid` text DEFAULT NULL,
  `table_info` json NOT NULL,
  `is_global` tinyint(1) DEFAULT NULL,
  `create_time` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
  `update_time` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  UNIQUE KEY `uk_id_schema_table` (`id`,`cp_schema`,`cp_table`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
1 row in set (0.0016 sec)
```

This is with DM v5.4.2.  I assume the unique keys with all columns defined as `NOT NULL` should be seen as good-enough.

bb7133 commented:
Hi @joycse06 , thanks for the contribution!

If this variable is supposed to be used with TiCDC-based replication, should we consider unique key as well?

I know that unique key is not included in `sql_require_primary_key` for MySQL, is it better to introduce another system variables for both primary key and unique key check?

morgo commented:
> If this variable is supposed to be used with TiCDC-based replication, should we consider unique key as well?

We don't have a use-case beyond what the MySQL feature offers. It's not a typical use-case to have a table which doesn't have a primary key, but has a unique key.

joycse06 commented on parser/parser.go:
```diff
@@ -26,9 +26,8 @@
 
```
These and the few other unrelated changes were done by my IDE (GoLand) automatically during a merge from upstream but they seem legit to me, so I checked them in.

joycse06 commented:
> If this variable is supposed to be used with TiCDC-based replication, should we consider unique key as well?

@bb7133 I agree with Morgan on this. As we don't have a compelling use case to go beyond what MySQL offers atm, I think we should have it do what MySQL does.

joycse06 commented on parser/parser.go:
```diff
@@ -26,9 +26,8 @@
-import __yyfmt__ "fmt"
-
 import (
+	__yyfmt__ "fmt"
```
These and the few other unrelated changes were done by my IDE (GoLand) automatically during a merge from upstream but they seem legit to me, so I checked them in.

morgo replied:
There are triggering the CI failure:

```
[2022-07-24T23:54:47.514Z]  
[2022-07-24T23:54:47.514Z] +import __yyfmt__ "fmt"
[2022-07-24T23:54:47.514Z] +
[2022-07-24T23:54:47.514Z]  import (
[2022-07-24T23:54:47.514Z] -	__yyfmt__ "fmt"
[2022-07-24T23:54:47.514Z]  	"strings"
```

Because it's generated code (from parser.y), it will be reverted by the next person :( So I think it is best to revert.

joycse06 replied:
Reverted these changes. It seems I had a pre-commit hook doing import optimisation.

bb7133 reviewed (approved):
LGTM, please fix the lint as commented by @morgo: https://github.com/pingcap/tidb/pull/36146#discussion_r930322136

bb7133 commented:
Hi @joycse06 , thanks for your contribution! Would you like to add the document for this variable(https://github.com/pingcap/docs/blob/master/system-variables.md)?

joycse06 commented:
@bb7133  Sure, I will give it a go. I will raise a PR in that repo when ready. Thanks.

joycse06 commented:
@bb7133  Can you please have a look at https://github.com/pingcap/docs/pull/9886 ? Thanks
